exercises/grep/description.md

```diff
@@ -1,77 +1,24 @@
 # Description
 
-Search a file for lines matching a regular expression pattern. Return the line
-number and contents of each matching line.
+Search files for lines matching a search string and return all matching lines.
 
-The Unix [`grep`](http://pubs.opengroup.org/onlinepubs/9699919799/utilities/grep.html) command can be used to search for lines in one or more files
-that match a user-provided search query (known as the *pattern*).
+The Unix [`grep`](http://pubs.opengroup.org/onlinepubs/9699919799/utilities/grep.html) command searches for lines in files that match a regular expression.
```

Comment:
@hkBst Did you dislike the use of "pattern" in the original description?

Reply:
I was not so happy that "pattern" was used ambiguously to refer both to "regular expression" and to "literal string", so I tried to clarify that while the real grep searches for regular expression( pattern)s, in this exercise you only need to implement literal string( pattern) search. The ambiguity of the word "pattern" made me try to avoid its use.

Reply:
I think the ambiguity is fine, since below it's clearly stated that this exercise is a simplification and only cares about literal strings. If people are unfamiliar with grep they can look at the linked documentation.

My vote would be to go with `pattern` here.

Reply:
And the fact that grep does match patterns, both with literal strings and regular expressions, it covers it (and reiterates what we find to be described in the first(?) line of the documentation.

Reply:
What exactly are you proposing? Do you just want to add the term "pattern"? Or do you want it to replace something else? And to what purpose?

The ambiguity about regular expression vs literal strings was perhaps my main motivation to take a look at improving the prose of this exercise. I am glad that you think this is now very clear, but also a bit afraid to lose this clarification.

Reply:
On further reflection I am happy to replace "regular expression" here with "pattern". Can you propose a change?

Reply:
I don't feel strongly about this at all, I was mostly just curious why you preferred regex to pattern when I had the exact opposite reaction :) Usually when this happens I learn something interesting from asking the question.

I have a very weak preference towards:

> The Unix [`grep`](http://pubs.opengroup.org/onlinepubs/9699919799/utilities/grep.html) command searches for lines in files that match a pattern.

But having had this discussion I'm perfectly happy with the current state of the sentence.

Reply:
This can be read as "files that match a regular expression" instead of "lines that match a regular expression"

Perhaps this (which hand-waves over grep processing its stdin, or handling multiple regexes):

> The grep command searches files for lines matching a regular expression.

I would also prefer "pattern" over "regular expression", mostly because `grep -F` exists to match fixed strings.

> The grep command searches files for lines matching a pattern.

Reply:
Ah good catch on that one.

> The grep command searches files for lines matching a pattern.

This is now my favorite.

Reply:
Since fixed strings are a special case of regular expressions, I don't think we gain anything by being non-specific with "pattern". If anyone wonders what a pattern is and they click through, they find out that it is basically a regular expression, although grep supports two kinds of regular expressions and there is special support for easily matching on fixed strings. So, therefore, I wonder why not just say regular expression, when that is basically all grep does?

I do agree that this word reordering is a definite improvement.

Reply:
One more data point: I think the wikipedia page for regular expression is much more relevant here than the one for pattern.

RE: A regular expression is a sequence of [characters](https://en.wikipedia.org/wiki/Character_(computing)) that specifies a [search pattern](https://en.wikipedia.org/wiki/Pattern_matching) in [text](https://en.wikipedia.org/wiki/String_(computer_science)).
Pattern: A pattern is a regularity in the world, in human-made design, or in [abstract](https://en.wikipedia.org/wiki/Abstraction) ideas.

Reply:
That works for me.